packages/webpack-extensive-lodash-replacement-plugin/index.js

```diff
@@ -442,18 +472,72 @@ class ExtensiveLodashReplacementPlugin {
 		return request;
 	}
 
+	async modifyResult( result ) {
+		if ( ! result || ! result.contextInfo || ! result.contextInfo.issuer ) {
+			return;
+		}
+
+		// Wait for initialization, if it's still running.
+		await this.init;
+
+		const { request } = result;
+
+		// Replace plain 'lodash' with 'lodash-es'.
+		if ( /^lodash$/.test( request ) ) {
+			if ( await this.canBeReplaced( result.contextInfo.issuer, 'lodash' ) ) {
+				result.request = 'lodash-es';
+				return;
+			}
+		}
+
+		// Replace 'lodash/foo' with 'lodash-es/foo'.
+		if ( /^lodash\/(.*)$/.test( request ) ) {
+			if ( await this.canBeReplaced( result.contextInfo.issuer, 'lodash' ) ) {
+				result.request = request.replace( 'lodash/', 'lodash-es/' );
+				return;
+			}
+		}
+
+		// Replace 'lodash.foo' with 'lodash-es/foo'.
+		if ( /^lodash\.(.*)$/.test( request ) ) {
+			if ( await this.canBeReplaced( result.contextInfo.issuer, request ) ) {
+				const match = /^lodash\.(.*)$/.exec( request );
+				let subModule = match[ 1 ];
+
+				// Normalize module names.
+				// This avoids code duplication due to module name case differences
+				// (e.g. 'camelcase' vs 'camelCase').
+				LODASH_MODULE_NAMES.forEach( casedModule => {
+					if ( subModule === casedModule.toLowerCase() ) {
+						subModule = casedModule;
+					}
+				} );
+
+				result.request = `lodash-es/${ subModule }`;
+				return;
+			}
+		}
+
+		return;
+	}
+
 	apply( compiler ) {
+		compiler.hooks.thisCompilation.tap( 'LodashReplacementPlugin', compilation => {
+			this.compilation = compilation;
+		} );
+
 		compiler.hooks.normalModuleFactory.tap( 'LodashReplacementPlugin', nmf => {
-			nmf.hooks.beforeResolve.tap( 'LodashReplacementPlugin', result => {
-				if ( ! result ) return;
-				result.request = this.getModifiedRequest( result );
-				return result;
-			} );
-			nmf.hooks.afterResolve.tap( 'LodashReplacementPlugin', result => {
-				if ( ! result ) return;
-				result.request = this.getModifiedRequest( result );
-				return result;
-			} );
+			this.moduleResolver = this.moduleResolver || nmf.getResolver( 'normal' );
+			this.init = this.init || this.initBaseLodashData();
+
+			nmf.hooks.beforeResolve.tapPromise(
+				'LodashReplacementPlugin',
+				this.modifyResult.bind( this )
+			);
+			nmf.hooks.afterResolve.tapPromise(
```

Comment:
After giving it some thought, I think that the `afterResolve` hook is not needed here. `NormalModuleReplacementPlugin` uses them both, and:
- the `beforeResolve` hook modifies `result.request`, which can be `lodash/get`, for example
- the `afterResolve` hook modifies `result.resource`, in case the request happened to resolve to something that matches the plugin's replacement rules. `resource` is a full path, e.g., `/Users/jsnajdr/src/wp-calypso/node_modules/lodash/get.js`

I think we only need to check the requests in `beforeResolve`, don't we? In any case, the current code that modifies `result.request` after the resolution has been done wouldn't have any effect anyway.

Reply:
I was following the `NormalModuleReplacementPlugin` codebase blindly here, as I've found it an enormous challenge to look for definitive documentation on webpack APIs. I've removed the `afterResolve` hook, let's see what happens :)